app/services/katello/pulp3/repository/yum.rb

```diff
@@ -59,9 +60,54 @@ def partial_repo_path
           "/pulp/repos/#{repo.relative_path}/".sub('//', '/')
         end
 
-        def copy_content_for_source
-          # TODO
-          fail NotImplementedError
+        def copy_units(source_repository_version_href, content_unit_hrefs, dependency_solving)
+          tasks = []
+
+          data = PulpRpmClient::Copy.new
+          data.config = [
+            source_repo_version: source_repository_version_href,
+            dest_repo: repository_reference.repository_href,
+            content: content_unit_hrefs
+          ]
+          data.dependency_solving = dependency_solving
+          tasks << api.copy_api.copy_content(data)
+
+          tasks
+        end
+
+        def copy_content_for_source(source_repository, options = {})
+          filters = ContentViewPackageFilter.where(:id => options[:filter_ids])
+          dependency_solving = options[:solve_dependencies] || false
+          whitelist_ids = []
+          blacklist_ids = []
+          filters.each do |filter|
+            if filter.inclusion
+              whitelist_ids += filter.content_unit_pulp_ids(source_repository)
+            else
+              blacklist_ids += filter.content_unit_pulp_ids(source_repository)
+            end
+          end
+
+          content_unit_hrefs = []
+          if whitelist_ids.empty?
+            content_unit_hrefs = source_repository.rpms.where(:modular => false).pluck(:pulp_id).sort - blacklist_ids
+          else
+            content_unit_hrefs = whitelist_ids - blacklist_ids
+          end
+
+          content_unit_hrefs += source_repository.srpms.pluck(:pulp_id)
+          errata_to_include = filter_by_pulp_href(source_repository.errata, content_unit_hrefs)
+          content_unit_hrefs += errata_to_include.collect { |erratum| erratum.repository_errata.pluck(:erratum_pulp3_href) }.flatten.uniq
+
+          copy_units(source_repository.version_href, content_unit_hrefs, dependency_solving)
+        end
+
+        def package_group(package_group_href, options = {})
+          api.content_package_groups_api.read(package_group_href, options)
+        end
```

Comment:
I think this may already be handled here: https://github.com/Katello/katello/blob/master/app/services/katello/pulp3/package_group.rb  or at least might be a better place to put it if not

Reply:
I don't even recall _why_ that exists!